Spigot-Server-Patches/0418-Ticket-level-per-player-view-distance-API.patch

```diff
@@ -0,0 +1,525 @@
++                    if(f == 1) {
++                        this.a(i, this.c(i), false, true, Integer.MAX_VALUE);
++                    } else if (f == -1) {
++                        this.a(i, this.c(i), true, false, Integer.MAX_VALUE);
++                    }
++                }
++
++                ChunkMapDistance.this.update.clear();
++                //Paper end
++
+                 this.g.clear();
+             }
+ 
+diff --git a/src/main/java/net/minecraft/server/EntityEnderDragon.java b/src/main/java/net/minecraft/server/EntityEnderDragon.java
+index c8c74f2b3..63f19c510 100644
+--- a/src/main/java/net/minecraft/server/EntityEnderDragon.java
++++ b/src/main/java/net/minecraft/server/EntityEnderDragon.java
+@@ -570,9 +570,9 @@ public class EntityEnderDragon extends EntityInsentient implements IMonster {
+                 // CraftBukkit start - Use relative location for far away sounds
+                 // this.world.b(1028, new BlockPosition(this), 0);
+                 // Paper start
+-                int viewDistance = ((WorldServer) this.world).spigotConfig.viewDistance * 16; // Paper - updated to use worlds actual view distance incase we have to uncomment this due to removal of player view distance API
++                //int viewDistance = ((WorldServer) this.world).spigotConfig.viewDistance * 16; // Paper - updated to use worlds actual view distance incase we have to uncomment this due to removal of player view distance API
+                 for (EntityPlayer player : ((WorldServer)world).getPlayers()) {
+-                    //final int viewDistance = player.getViewDistance(); // TODO apply view distance api patch
++                    final int viewDistance = player.getTrackViewDistance(); //Paper ticket level pp-view-dist
+                     // Paper end
+                     double deltaX = this.locX - player.locX;
+                     double deltaZ = this.locZ - player.locZ;
+diff --git a/src/main/java/net/minecraft/server/EntityHuman.java b/src/main/java/net/minecraft/server/EntityHuman.java
+index cd73cde8d..809edf46f 100644
+--- a/src/main/java/net/minecraft/server/EntityHuman.java
++++ b/src/main/java/net/minecraft/server/EntityHuman.java
+@@ -80,6 +80,29 @@ public abstract class EntityHuman extends EntityLiving {
+     public String spawnWorld = "";
+     public int oldLevel = -1;
+ 
++    //Paper start ticket level pp-view-dist
++    private int viewDistance;
++    private int trackViewDistance;
++    public int lastViewDistance;
++    public int lastTrackViewDistance;
++    public int getViewDistance() { return this.viewDistance; }
++    public int getTrackViewDistance() { return this.trackViewDistance; }
++    public void setTrackViewDistance(int i) {
++        this.lastTrackViewDistance = this.trackViewDistance;
++        this.trackViewDistance = MathHelper.clamp(i, 3, 33);
++        if(i != this.lastTrackViewDistance) {
++            if (this.tracker != null) this.tracker.updatePlayer((EntityPlayer) this);
```

Comment:
this is not how you propagate tracking range changes

Reply:
correct method?

Reply:
as a note this has been moved down to viewDistance() due to your last comment

Reply:
in general leave this out, it gets handled next tick

Reply:
will be committed